com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java

```diff
@@ -348,19 +378,29 @@ public static String getDefaultDmengineExeName(String variant) {
         }
     }
 
-    public static List<String> getDefaultDmenginePaths(Platform platform, String variant) throws IOException {
+    private static List<String> getDefaultDmenginePaths(Platform platform, String variant) throws IOException {
         return getExes(platform, getDefaultDmengineExeName(variant));
     }
 
-    public static List<File> getDefaultDmengineFiles(Platform platform, String variant) throws IOException {
-        List<String> binaryPaths = getDefaultDmenginePaths(platform, variant);
-        List<File> binaryFiles = new ArrayList<File>();
-        for (String path : binaryPaths) {
-            binaryFiles.add(new File(path));
+    public static List<File> getDefaultDmengineFiles(Platform platform, String variant, String artifactsURL) throws IOException {
+        List<File> binaryFiles;
+        try {
+            List<String> binaryPaths = getDefaultDmenginePaths(platform, variant);
+            binaryFiles = new ArrayList<File>();
+            for (String path : binaryPaths) {
+                binaryFiles.add(new File(path));
+            }
+        }
+        catch (RuntimeException e) {
+            binaryFiles = downloadExes(platform, variant, artifactsURL);
```

Comment:
Instead of doing it only for headless build, it makes much more sense to check if executable included in bob, if not try to download it, and if file doesn't exist then report an error